jtwig-core/src/main/java/org/jtwig/Environment.java

```diff
@@ -0,0 +1,279 @@
+        
+        try {
+            ReportingParseRunner<Compilable> runner = new ReportingParseRunner<>(parser.start());
+            ParsingResult<Compilable> result = runner.run(
+                    FileUtils.readAllText(resource.source(), Charset.defaultCharset()));
+            cached = (Template)result.resultValue;
+            getCache().addParsed(resource.getCacheKey(), cached);
+            return cached;
+        } catch (ParserRuntimeException e) {
+            if (e.getCause() instanceof ParseBypassException) {
+                ParseException innerException = ((ParseBypassException) e.getCause()).getInnerException();
+                innerException.setExpression(e.getMessage());
+                throw innerException;
+            } else {
+                throw new ParseException(e);
+            }
+        } catch (ResourceException e) {
+            throw new ParseException(e);
+        }
+    }
+    
+    public Template.CompiledTemplate compile(final String name)
+            throws ResourceException, ParseException, CompileException {
+        Loader.Resource resource = load(name);
+        Template template = parse(resource);
+        return compile(template, resource);
+    }
+    public Template.CompiledTemplate compile(final String name,
+            final CompileContext context)
+            throws ResourceException, ParseException, CompileException {
+        Loader.Resource resource = load(name);
+        Template template = parse(resource);
+        return compile(template, resource, context);
+    }
+    public Template.CompiledTemplate compile(final Loader.Resource resource)
+            throws ResourceException, ParseException, CompileException {
+        Template template = parse(resource);
+        return compile(template, resource);
+    }
+    public Template.CompiledTemplate compile(final Loader.Resource resource,
+            final CompileContext context)
+            throws ResourceException, ParseException, CompileException {
+        Template template = parse(resource);
+        return compile(template, resource);
+    }
+    public Template.CompiledTemplate compile(final Template template,
+            final Loader.Resource resource)
+            throws ResourceException, ParseException, CompileException {
+        CompileContext compileContext = new CompileContext(resource, this);
+        return compile(template, resource, compileContext);
+    }
+    public Template.CompiledTemplate compile(final Template template,
+            final Loader.Resource resource,
+            final CompileContext context) throws CompileException {
+        Template.CompiledTemplate cached = getCache().getCompiled(resource.getCacheKey());
+        if (cached != null) {
+            return cached;
+        }
+        cached = template.compile(context);
+        getCache().addCompiled(resource.getCacheKey(), cached);
+        return cached;
+    }
+}
```

Comment:
This is not very important, it's always a tradeoff. Having a class touching many so different contexts might lead us to what is called as "Divergent Change" code smell. This environment class is doing a lot of different things:

- Holding the configuration
- Parse/Compiling logic

I would suggest to go with a separation of duties and place the configurations on a proper class and use a message chain to get/set them.